poolakey/src/main/java/ir/cafebazaar/poolakey/billing/connection/ReceiverBillingConnection.kt

```diff
@@ -0,0 +1,411 @@
+        purchaseType: PurchaseType,
+        callback: PurchaseIntentCallback.() -> Unit
+    ) {
+        purchaseActivityWeakReference = WeakReference(
+            PurchaseWeakHolder(activity, purchaseRequest.requestCode, callback)
+        )
+
+        sendPurchaseBroadcast(purchaseRequest, purchaseType, callback)
+    }
+
+    override fun purchase(
+        fragment: Fragment,
+        purchaseRequest: PurchaseRequest,
+        purchaseType: PurchaseType,
+        callback: PurchaseIntentCallback.() -> Unit
+    ) {
+        purchaseFragmentWeakReference = WeakReference(
+            PurchaseWeakHolder(fragment, purchaseRequest.requestCode, callback)
+        )
+
+        sendPurchaseBroadcast(purchaseRequest, purchaseType, callback)
+    }
+
+    private fun sendPurchaseBroadcast(
+        purchaseRequest: PurchaseRequest,
+        purchaseType: PurchaseType,
+        callback: PurchaseIntentCallback.() -> Unit
+    ) {
+        PurchaseIntentCallback().apply(callback).purchaseFlowBegan.invoke()
+        getNewIntentForBroadcast().apply {
+            action = ACTION_PURCHASE
+            putExtra(KEY_SKU, purchaseRequest.productId)
+            putExtra(KEY_DEVELOPER_PAYLOAD, purchaseRequest.payload)
+            putExtra(KEY_ITEM_TYPE, purchaseType.type)
+
+        }.run {
+            sendBroadcast(this)
+        }
+    }
+
+
+    override fun stopConnection() {
+        disconnected = true
+
+        cleanReferences()
+
+        receiverCommunicator?.let {
+            BillingReceiver.removeObserver(it)
+        }
+        receiverCommunicator = null
+    }
+
+    private fun cleanReferences() {
+        consumeCallback = null
+        queryCallback = null
+        connectionCallbackReference = null
+        contextReference = null
+
+        purchaseFragmentWeakReference = null
+        purchaseActivityWeakReference = null
+    }
+
+    private fun onQueryPurchaseReceived(extras: Bundle?) {
+        queryCallback?.let {
+            queryFunction.function(
+                QueryFunctionRequest(
+                    "",
+                    { _, _ -> extras },
+                    it
+                )
+            )
+        }
+    }
+
+    private fun onPurchaseReceived(extras: Bundle?) {
+        if (isResponseSucceed(extras)) {
+            when {
+                purchaseActivityWeakReference?.get() != null -> {
+                    startPurchaseActivityWithActivity(
+                        requireNotNull(purchaseActivityWeakReference?.get()),
+                        getPurchaseIntent(extras)
+                    )
+                }
+                purchaseFragmentWeakReference?.get() != null -> {
+                    startPurchaseActivityWithFragment(
+                        requireNotNull(purchaseFragmentWeakReference?.get()),
+                        getPurchaseIntent(extras)
+                    )
+                }
+                else -> {
+                    // TODO handle this
+                }
```

Comment:
What's preventing us from handling this case?

Reply:
I've changed it to comment, as all callback of purchase is null, we can't invoke it, connection disconnecting is only available state